src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -205,80 +205,78 @@ ValidatorList::applyList (
         TimeKeeper::duration{list["expiration"].asUInt()}};
-    std::vector<PublicKey>& publisherList = publisherLists_[pubKey].list;
+    auto& publisherList = publisherLists_[pubKey].list;
 
-    std::vector<PublicKey> oldList = publisherList;
-    publisherList.clear ();
-    publisherList.reserve (newList.size ());
+    auto oldList = publisherList;
+    publisherList.clear();
+    publisherList.reserve(newList.size());
     std::vector<std::string> manifests;
     for (auto const& val : newList)
     {
-        if (val.isObject() &&
-            val.isMember ("validation_public_key") &&
-            val["validation_public_key"].isString ())
+        if (val.isObject())
         {
-            std::pair<Blob, bool> ret (strUnHex (
-                val["validation_public_key"].asString ()));
-
-            if (! ret.second || ! publicKeyType(makeSlice(ret.first)))
-            {
-                JLOG (j_.error()) <<
-                    "Invalid node identity: " <<
-                    val["validation_public_key"].asString ();
-            }
-            else
+            if (val.isMember("validation_public_key") &&
+                val["validation_public_key"].isString())
             {
-                publisherList.push_back (
-                    PublicKey(Slice{ ret.first.data (), ret.first.size() }));
+                std::pair<Blob, bool> ret(
+                    strUnHex(val["validation_public_key"].asString()));
+
+                if (!ret.second || !publicKeyType(makeSlice(ret.first)))
+                {
+                    JLOG (j_.error()) << "Invalid node identity: "
+                        << val["validation_public_key"].asString();
+                    continue;
+                }
+
+                publisherList.emplace_back(
+                    PublicKey{Slice{ret.first.data(), ret.first.size()}});
             }
 
-            if (val.isMember ("manifest") && val["manifest"].isString ())
-                manifests.push_back(val["manifest"].asString ());
+            if (val.isMember("manifest") && val["manifest"].isString())
+                manifests.push_back(val["manifest"].asString());
         }
     }
 
-    // Update keyListings_ for added and removed keys
-    std::sort (
-        publisherList.begin (),
-        publisherList.end ());
-
-    auto iNew = publisherList.begin ();
-    auto iOld = oldList.begin ();
-    while (iNew != publisherList.end () ||
-        iOld != oldList.end ())
     {
-        if (iOld == oldList.end () ||
-            (iNew != publisherList.end () &&
-            *iNew < *iOld))
+        // Update keyListings_ for added and removed keys
+        std::sort(publisherList.begin(), publisherList.end());
+
         {
-            // Increment list count for added keys
-            ++keyListings_[*iNew];
-            ++iNew;
+            // Removed elements appear in the old list but not in the new
+            std::vector<PublicKey> removed;
```

Comment:
This new code creates 2 `std::vectors` and copies a bunch of elements into each (with `set_difference`), set_difference also iterates over the lists twice. The old code has better performance, albeit uglier/harder to read. Would there be a way to keep the existing algorithm but add comments to explain it better?